text/0000-fix-error.md

```diff
@@ -0,0 +1,246 @@
+
+```rust
+trait Error: Display + Debug {
+    fn backtrace(&self) -> Option<&Backtrace> {
+        None
+    }
+
+    fn source(&self) -> Option<&dyn Error + 'static> {
+        None
+    }
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The new API of the Error trait
+
+The new API provides three main components:
+
+1. The Display and Debug impls, for printing error messages. Ideally, the
+   Display API would be focused on *end user* messages, whereas the Debug impl
+   would contain information relevant to the programmer.
+2. The backtrace method. If the Error contains a backtrace, it should be exposed
+   through this method. Errors are not required to contain a backtrace, and are
+   not expected to.
+3. The `source` method. This returns another Error type, which is the underlying
+   source of this error. If this error has no underlying source (that is, it is
+   the "root source" of the error), this method should return none.
+
+## The backtrace API
+
+This RFC adds a new `backtrace` module to std, with one type, with this API:
+
+```rust
+pub struct Backtrace {
+    // ...
+}
+
+impl Backtrace {
+    pub fn new() -> Backtrace {
+        // ...
+    }
+}
+
+impl Display for Backtrace {
+    // ...
+}
+
+impl Debug for Backtrace {
+    // ...
+}
+```
+
+This minimal initial API is just intended for printing backtraces for end users.
+In time, this may grow the ability to visit individual frames of the backtrace.
+
+### Backtrace controlling environment variables
+
+Today, the `RUST_BACKTRACE` controls backtraces generated by panics. After this
+RFC, it also controls backtraces generated in the standard library: no backtrace
+will be generated when calling `Backtrace::new` unless this variable is set.
+
+Two additional variables will be added: `RUST_PANIC_BACKTRACE` and
+`RUST_STD_BACKTRACE`: these will independently override the behavior of
+`RUST_BACKTRACE` for backtraces generated for panics and from the std API.
+
+## The transition plan
+
+Deprecating both `cause` and `description` is a backward compatible change, and
+adding provided methods `backtrace` and `source` is also backward compatible.
+We can make these changes unceremoniously, and the `Error` trait will be much
+more functional.
+
+We also change the default definition of `cause`, even though it is deprecated:
+
+```rust
+fn cause(&self) -> Option<&dyn Error> {
+    self.source()
+}
+```
+
+This way, if an error defines `source`, someone calling the deprecated `cause`
+API will still get the correct cause type, even though they can't downcast it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why `cause` -> `source`
+
+The problem with the existing `cause` API is that the error it returns is not
+`'static`. This means it is not possible to downcast the error trait object,
+because downcasting can only be done on `'static` trait objects (for soundness
+reasons).
```

Comment:
Can you explain those soundness reasons, or at least link to an explanation? As I understand it,`Any` requires `'static` because otherwise you would be able to use the downcast methods to increase the lifetime of things... and the reason those downcast methods couldn't be altered to work for any lifetime `'a` is because there is no way in current Rust to require that a type `T` does not live longer that `'a`. Sounds like we could use another RFC...

Reply:
Lifetimes are erased so they aren't included in the typeid, so `Foo<'a>` and `Foo<'b>` always have the same typeid regardless of the relationship between those lifetimes.